File: packages/kotlin/src/commonMain/kotlin/io/github/rtarik/paintmixer/PipelinePaintMixer.kt
package io.github.rtarik.paintmixer

class PipelinePaintMixer(
    val baseEngine: BaseMixEngine,
    val correctionModel: ResidualCorrectionModel? = null,
) : PaintMixer {
    override fun mixOrNull(portions: List<MixPortion>): SrgbColor? {
        val baseMix = baseEngine.mixOrNull(portions) ?: return null
        if (correctionModel == null || portions.size <= 1) return baseMix
        return correctionModel.correct(portions, baseMix)
    }
}

File: packages/js/src/pipeline-paint-mixer.js
export class PipelinePaintMixer {
  constructor({ baseEngine, correctionModel = null, enforceCompatibility = true }) {
    if (!baseEngine || typeof baseEngine.mixOrNull !== 'function') {
      throw new TypeError('baseEngine must expose mixOrNull(portions)');
    }
    this.baseEngine = baseEngine;
    this.correctionModel = correctionModel;
    this.enforceCompatibility = enforceCompatibility;
  }

  mixOrNull(portions) {
    const baseMix = this.baseEngine.mixOrNull(portions);
    if (baseMix == null) return null;
    if (this.correctionModel == null || portions.length <= 1) return baseMix;

    const expectedBaseEngineId = this.correctionModel.expectedBaseEngineId;
    if (
      this.enforceCompatibility &&
      expectedBaseEngineId != null &&
      expectedBaseEngineId !== this.baseEngine.engineId
    ) {
      throw new Error(
        `Correction model ${this.correctionModel.modelId} expects base engine ${expectedBaseEngineId}, got ${this.baseEngine.engineId}`,
      );
    }

    return this.correctionModel.correct(portions, baseMix);
  }
}
